### PR TITLE
Add moule-nft

### LINKED ISSUE #416: module-nft
Depends on https://github.com/open-web3-stack/open-runtime-module-library/issues/290

  - Types
    - struct ClassData
      - deposit: Balance
      - properties: ClassProperties
    - enum flag ClassProperty
      - Transferable
      - Burnable
    - struct TokenData
      - deposit: Balance
- Trait: orml_nft::Trait + pallet_proxy::Trait
  - type CreateClassDeposit: Get\<Balance>
  - type CreateTokenDeposit: Get\<Balance>
  - type ConvertClassData: IsType\<Self::ClassData> +IsType\<ClassData>
  - type ConvertTokenData: IsType\<Self::TokenData> +IsType\<TokenData>
  - type TypeId: TypeId
  - type Currency: BasicReservableCurrency
- Storages
- Calls
  - fn create_class(origin, metadata: CID, properties: ClassProperties)
    - let next_id = // get next class id
    - let owner = T::TypeId::into_sub_account(next_id)
    - let deposit = CreateClassDeposit::get()
    - T::Currency::transfer(origin, owner, deposit)
    - T::Currency::reserve(owner, deposit)
    - pallet_proxy::Module\<T>::add_proxy(owner, origin, Default::default(), 0)
      - Depends on https://github.com/paritytech/substrate/issues/7139
      - For now, use origin as owner and skip the proxy part
    - orml_nft::Module\<T>::create_class(owner, metadata, ClassData { deposit, properties })
  - fn mint(origin, to: AccountId, class: ClassId, metadata: CID, quantity: u32)
    - ensure!(origin == Self::classes(class).owner)
    - let deposit = CreateTokenDeposit::get() * quantity
    - T::Currency::reserve(origin, deposit)
    - for _ in 0..quantity
      - orml_nft::Module\<T>::mint(to, metadata, TokenData { deposit })
  - fn transfer(origin, to: AccountId, token: (ClassId, TokenId))
    - check if transferable and transfer
  - fn burn(origin, token: (ClassId, TokenId))
    - check if burnable and burn
    - unreserve deposit and send to origin
  - fn destroy_class(origin, class_id: ClassId, dest: AccountId)
    - let info = Self::classes(class_id)
    - ensure!(orign == info.owner)
    - ensure!(info.total_issuance == 0)
    - unreserve deposit 
    - pallet_proxy::Module\<T>::remove_proxies(owner)
    - transfer all free from origin to `dest`
      - Skip above two steps until pallet_proxy is accessable

runtime

 - Types
    - struct NftTypeId
       - impl TypeId
          - TYPE_ID = b"aNFT"

### PR DESCRIPTION
**Pull request by zjb0807**
Closes #416

### CODE REVIEW

File: modules/nft/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	traits::{AccountIdConversion, Zero},
+	ModuleId, RuntimeDebug,
+};
+use sp_std::vec::Vec;
+
+mod mock;
+mod tests;
+
+#[derive(Encode, Decode, Clone, RuntimeDebug, PartialEq, Eq)]
+pub enum ClassProperty {
```
Comment (xlc) [inline modules/nft/src/lib.rs:23]: use `use enumflags2::BitFlags` to make this a bit flags

example:
https://github.com/paritytech/substrate/blob/596f377d49fede088e0c32791fb72f1343003b0d/frame/identity/src/lib.rs#L281-L311

File: modules/nft/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+				ensure!(properties.len() <= 2, Error::<T>::InvalidPropertiesLength);
+				let next_id = orml_nft::Module::<T>::next_class_id();
+				let owner: T::AccountId = T::ModuleId::get().into_sub_account(next_id);
+				let deposit = T::CreateClassDeposit::get();
+				<T as Trait>::Currency::transfer(&who, &owner, deposit)?;
+				<T as Trait>::Currency::reserve(&owner, deposit)?;
+				//	Depends on https://github.com/paritytech/substrate/issues/7139
+				//	For now, use origin as owner and skip the proxy part
+				//	pallet_proxy::Module<T>::add_proxy(owner, origin, Default::default(), 0)
+				let data = ClassData{deposit, properties};
```
Comment (xlc) [inline modules/nft/src/lib.rs:128]:
```suggestion
				let data = ClassData { deposit, properties };
```

File: modules/nft/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+				ensure!(data.into().properties.contains(&ClassProperty::Burnable), Error::<T>::NonBurnable);
+
+				let token_info = orml_nft::Module::<T>::tokens(token.0, token.1).ok_or(Error::<T>::TokenIdNotFound)?;
+				ensure!(who == token_info.owner, Error::<T>::NoPermission);
+
+				orml_nft::Module::<T>::burn(&who, token)?;
+				let owner: T::AccountId = T::ModuleId::get().into_sub_account(token.0);
+				let data: T::ConvertTokenData = From::from(token_info.data);
+				let data: TokenData = data.into();
+				<T as Trait>::Currency::unreserve(&owner, data.deposit);
```
Comment (xlc) [inline modules/nft/src/lib.rs:195]: use `repatriate_reserved` to unreserve and transfer at same time

File: modules/nft/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+				let who = ensure_signed(origin)?;
+				let class_info = orml_nft::Module::<T>::classes(class_id).ok_or(Error::<T>::ClassIdNotFound)?;
+				ensure!(who == class_info.owner, Error::<T>::NoPermission);
+				ensure!(class_info.total_issuance == Zero::zero(), Error::<T>::CannotDestroyClass);
+
+				let owner: T::AccountId = T::ModuleId::get().into_sub_account(class_id);
+				let data: T::ConvertClassData = From::from(class_info.data);
+				let data: ClassData = data.into();
+				// reserve balance should equal CreateClassDeposit.
+				ensure!(data.deposit == <T as Trait>::Currency::reserved_balance(&owner), Error::<T>::CannotDestroyClass);
```
Comment (xlc) [inline modules/nft/src/lib.rs:215]: I don't think we want this check. owner may have reserved balances from other pallets.

File: runtime/dev/src/lib.rs [hunk 5]
```diff
@@ -1307,6 +1359,7 @@ construct_runtime!(
 		NomineesElection: module_nominees_election::{Module, Call, Storage},
 		StakingPool: module_staking_pool::{Module, Call, Storage, Event<T>},
 		PolkadotBridge: module_polkadot_bridge::{Module, Call, Storage, Event<T>, Config},
+		AcalaNFT: module_nft::{Module, Call, Event<T>},
```
Comment (xlc) [inline runtime/dev/src/lib.rs:1362]:
```suggestion
		NFT: module_nft::{Module, Call, Event<T>},
```

File: modules/nft/src/lib.rs (hunk 3 above)
Comment (zjb0807) [reply modules/nft/src/lib.rs:195]: https://github.com/paritytech/substrate/blob/master/frame/balances/src/lib.rs#L1263
If `to` account is not exist, `repatriate_reserved` will return `DeadAccount`.
But `transfer` did not do this check.

Comment (xlc) [reply modules/nft/src/lib.rs:210]: ok. you can add that to comment so next reviewer won't ask the same question.

**Review by xlc — APPROVED** [review]

**Pull request merged**